### PR TITLE
fix: allow deletion of user when it's order are being past

### PR DESCRIPTION
**Pull request by maze-runnar**
Fixes https://github.com/fossasia/open-event-frontend/pull/6501

#### Checklist

- [x] I have read the [Contribution & Best practices Guide](https://blog.fossasia.org/open-source-developer-guide-and-best-practices-at-fossasia) and my PR follows them.
- [x] My branch is up-to-date with the Upstream `development` branch.
- [ ] The unit tests pass locally with my changes 
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

- [ ] All the functions created/modified in this PR contain relevant docstrings.

### CODE REVIEW

File: app/api/users.py [hunk 1]
```diff
@@ -309,8 +310,15 @@ def before_update_object(self, user, data, view_kwargs):
...
+
+                    present_event_orders = db.session.query(
+                        TicketHolder.query.filter_by(user=user)
+                        .join(Order.event)
+                        .filter(Event.ends_at > datetime.now())
+                        .exists()
+                    ).scalar()
                     # If any pending or completed order exists, we cannot delete the user
-                    if order_exists:
+                    if order_exists and present_event_orders:
```
Comment (iamareebjamal) [inline app/api/users.py:321]: Doesn't fix anything

**Comment by iamareebjamal** [comment]
How does this fix the issue. Did you test it?

**Comment by maze-runnar** [comment]
> How does this fix the issue. Did you test it?

yes I have tested it. It is working.

**Comment by iamareebjamal** [comment]
What does it fix? I don't understand

The issue is about not preventing account deletion if the tickets are in past, but here, instead you add a new check without modifying the old check. How can it fix the issue?

**Comment by maze-runnar** [comment]
> What does it fix? I don't understand
> 
> The issue is about not preventing account deletion if the tickets are in past, but here, instead you add a new check without modifying the old check. How can it fix the issue?

So i have added a check with  `and`. if order -> event -> endsAt is greater than present time is true and status of ticket is not cancelled is true(if and only if both are true) it will throw forbidden error. otherwise deletion will happen as usual.

**Comment by iamareebjamal** [comment]
Why not remove the old check and only check for new check. The old check is redundant

**Comment by iamareebjamal** [comment]
Change this filter in old query itself. No need for 2 queries. More chances of bugs

**Comment by maze-runnar** [comment]
old check is checking status of order, and this one is event end time of that order. how any of it is redundant?

**Comment by maze-runnar** [comment]
tested" 
Screenshot from 2021-01-29 00-16-32

File: app/api/users.py [hunk 2]
```diff
@@ -299,16 +300,21 @@ def before_update_object(self, user, data, view_kwargs):
                     order_exists = db.session.query(
                         TicketHolder.query.filter_by(user=user)
                         .join(Order)
+                        .join(Order.event)
```
Comment (iamareebjamal) [inline app/api/users.py:303]:
```suggestion
                        .join(Order.event)
                        .filter(Event.ends_at > datetime.now())
```

File: app/api/users.py [hunk 3]
```diff
@@ -299,16 +300,21 @@ def before_update_object(self, user, data, view_kwargs):
...
-                                Order.status == 'initializing',
-                                Order.status == 'pending',
+                            and_(
+                                or_(
+                                    Order.status == 'completed',
+                                    Order.status == 'placed',
+                                    Order.status == 'initializing',
+                                    Order.status == 'pending',
+                                ),
+                                Event.ends_at > datetime.now(),
```
Comment (iamareebjamal) [inline app/api/users.py:312]:
```suggestion
                                or_(
                                    Order.status == 'completed',
                                    Order.status == 'placed',
                                    Order.status == 'initializing',
                                    Order.status == 'pending',
```

File: app/api/users.py [hunk 4]
```diff
@@ -299,16 +300,21 @@ def before_update_object(self, user, data, view_kwargs):
...
                             )
                         )
                         .exists()
                     ).scalar()
+
```
Comment (iamareebjamal) [inline app/api/users.py:317]:
```suggestion
```

File: app/api/users.py [hunk 5]
```diff
@@ -1,10 +1,11 @@
 import base64
 import logging
+from datetime import datetime
 
 from flask import Blueprint, abort, jsonify, make_response, render_template, request
 from flask_jwt_extended import current_user, verify_fresh_jwt_in_request
 from flask_rest_jsonapi import ResourceDetail, ResourceList, ResourceRelationship
-from sqlalchemy import or_
+from sqlalchemy import and_, or_
```
Comment (iamareebjamal) [inline app/api/users.py:8]:
```suggestion
from sqlalchemy import or_
```

File: app/api/users.py [hunk 6]
```diff
@@ -299,16 +300,19 @@ def before_update_object(self, user, data, view_kwargs):
...
-                            or_(
-                                Order.status == 'completed',
-                                Order.status == 'placed',
-                                Order.status == 'initializing',
-                                Order.status == 'pending',
+                              or_(
+                                  Order.status == 'completed',
+                                  Order.status == 'placed',
+                                  Order.status == 'initializing',
+                                  Order.status == 'pending',
```
Comment (iamareebjamal) [inline app/api/users.py:310]:
```suggestion
                            or_(
                                Order.status == 'completed',
                                Order.status == 'placed',
                                Order.status == 'initializing',
                                Order.status == 'pending',
```

**Comment by iamareebjamal** [comment]
Check the solution now

**Comment by iamareebjamal** [comment]
Is it working?

**Comment by maze-runnar** [comment]
> Is it working?

yes

**Pull request merged**